rfd/0227-ec2-discovery-by-aws-org.md

```diff
@@ -0,0 +1,442 @@
+This API returns all the Accounts under the current organization.
+
+Using the Assume Role feature, which already exists, the service will assume a role in each AWS Account and run the same process it has today.
+
+```go
+func (f *ec2InstanceFetcher) GetInstances(ctx context.Context, rotation bool) ([]Instances, error) {
+  // ...
+  var assumeRoles []*types.AssumeRole
+  accountList, err := organizationsClient.ListAccounts(ctx, &organizations.ListAccountsInput{})
+  for _, account := range accountList.Accounts {
+    assumeRoleARNParts.AccountID = aws.ToString(account.Id)
+
+    assumeRoles = append(assumeRoles, &types.AssumeRole{
+      RoleARN:    assumeRoleARNParts.String(),
+      ExternalID: f.AssumeRole.ExternalID,
+    })
+  }
+
+  // ...
+  for _, assumeRole := range assumeRoles {
+    ec2Client, err := f.EC2Getter(ctx, f.Region, assumeRole, f.AWSClientsOpts...)
+    if err != nil {
+      return nil, trace.Wrap(err)
+    }
+
+    // use ec2Client to call ec2.DescribeInstances and submit instances
+    // afterwards, ssm.SendCommand will be used to install teleport into them
+  }
+}
+```
+
+### Teleport IAM Join based on AWS Organization
+
+In order for an instance to join Teleport using the IAM Join method, the instance signs an `sts.GetCallerIdentity` request and sends it to the Auth service.
+The Auth executes the request on behalf of the instance.
+The response contains the Account ID, and that's how Teleport ensures only instances whose account id is the same as the IAM Join Token, are accepted.
+
+This RFD aims to add a small change to this flow: allow validation to be based on the AWS Organization.
+
+When an EC2 instance tries to join the cluster using the IAM method, it will:
+- create a signed request for `sts.GetCallerIdentity`, which already happens today
+- create a signed request for `organizations.DescribeOrganization`
+- send both signed requests
+
+The Auth server might not execute both requests.
+If the IAM Join token has any validation by account id or assumed role, then it will execute the `sts.GetCallerIdentity`.
+If the token validates the AWS Organization ID, then it will execute the `organizations.DescribeOrganization`.
+
+This prevents unnecessary API calls.
+
+The `sts.GetCallerIdentity` API call should not fail, for well-behaved clients: there's not specific IAM permissions required, and clients only send it after doing some validation.
+
+The `organizations.DescribeOrganization` API call might fail, even when the client sends it.
+The client does not perform the API calls before signing them, so they might not know that they lack permissions.
+
+Contrary to the `sts.GetCallerIdentity` - which doesn't require any specific permission -, the `organizations.DescribeOrganization` requires that permission to be allowed by the EC2's assigned IAM Role.
```

Comment:
This will not be needed if we do this on the control plane side like mentioned above.